### PR TITLE
Git 2.28 no longer permits diff with ... on unrelated branches

### PR DESCRIPTION
**Pull request by zeripath**
Annoyingly git 2.28 now strictly enforces that when doing `git diff base...head` base and head must have a merge-base. Unfortunately this is a breaking change as previously this would just fall back to `git diff base head`. Therefore we need to detect this new error and handle it. 

In preparation for further changes I have also predicted that rev-list base...head may break in future too. Therefore I have attempted to protect against this too.

1.12 will also be affected by this and therefore we should backport these changes if possible. 

Closes #12362

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

**Comment by 6543** [comment]
```
Internal Server Error in 132.361604ms
    pull_merge_test.go:317: 
                Error Trace:    integration_test.go:411
                                                        integration_test.go:265
                                                        pull_merge_test.go:317
                                                        git_helper_for_declarative_test.go:108
                                                        pull_merge_test.go:253
                Error:          Not equal: 
                                expected: 201
                                actual  : 500
                Test:           TestCantMergeUnrelated
                Messages:       Request: POST /api/v1/repos/user1/repo1/pulls?token=e1040e3a70e06cfa25117f810803e9b1210f0eb1
    pull_merge_test.go:317: Response: {"message":"","url":"http://localhost:3003/api/swagger"}
    unit_tests.go:173: 
                Error Trace:    unit_tests.go:173
                                                        pull_merge_test.go:322
                                                        git_helper_for_declarative_test.go:108
                                                        pull_merge_test.go:253
                Error:          Should be true
                Test:           TestCantMergeUnrelated
                Messages:       Expected to find &{ID:0 Type:0 Status:0 ConflictedFiles:[] CommitsAhead:0 CommitsBehind:0 IssueID:0 Issue:<nil> Index:0 HeadRepoID:99 HeadRepo:<nil> BaseRepoID:99 BaseRepo:<nil> HeadBranch:unrelated BaseBranch:base ProtectedBranch:<nil> MergeBase: HasMerged:false MergedCommitID: MergerID:0 Merger:<nil> MergedUnix:0 isHeadRepoLoaded:false} (of type *models.PullRequest, with conditions []), but did not
```

File: modules/git/repo_compare.go [hunk 1]
```diff
@@ -115,14 +115,27 @@ func (repo *Repository) GetDiffNumChangedFiles(base, head string) (int, error) {
 
 	if err := NewCommand("diff", "-z", "--name-only", base+"..."+head).
 		RunInDirPipeline(repo.Path, w, stderr); err != nil {
+		if strings.Contains(err.Error(), "no merge base") {
```
Comment (zeripath) [inline modules/git/repo_compare.go]:
```suggestion
		if strings.Contains(stderr.String(), "no merge base") {
```

**Comment by zeripath** [comment]
OK try again - I missed checking the stderr properly

**Review by 6543 — APPROVED** [review]
🚀

**Review by techknowlogick — APPROVED** [review]

**Review by 6543 — CHANGES_REQUESTED** [review]
block - patch failed - investigate ...

**Review by 6543 — APPROVED** [review]
ok 🚀 (it had nothing to do with the git 2.28 problem)

**Pull request merged**